### ISSUE TITLE
[MJAVADOC-504] NullPointerException in JavadocUtil.getJavaHome() when JAVA_HOME isn't set

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#796]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** opened **[MJAVADOC-504](https://issues.apache.org/jira/browse/MJAVADOC-504?redirect=false)** and commented

Hi,

After upgrading to maven-javadoc-plugin I noticed a NullPointerException in `JavadocUtil.getJavaHome()` when the JAVA_HOME environment variable isn't set:

```
java.lang.NullPointerException
    at java.io.File.<init> (File.java:277)
    at org.apache.maven.plugins.javadoc.JavadocUtil.getJavaHome (JavadocUtil.java:1302)
    at org.apache.maven.plugins.javadoc.JavadocUtil.invokeMaven (JavadocUtil.java:941)
    at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.getModulesLinks (AbstractJavadocMojo.java:5640)
    at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addLinkofflineArguments (AbstractJavadocMojo.java:4046)
    at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.addStandardDocletOptions (AbstractJavadocMojo.java:4936)
    at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.executeReport (AbstractJavadocMojo.java:2074)
    at org.apache.maven.plugins.javadoc.JavadocReport.generate (JavadocReport.java:134)
    at org.apache.maven.plugins.javadoc.JavadocReport.doExecute (JavadocReport.java:329)
    at org.apache.maven.plugins.javadoc.AbstractJavadocMojo.execute (AbstractJavadocMojo.java:1909)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:993)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:345)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:191)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:180)
```

This error doesn't occur with the version 3.0.0-M1 or when JAVA_HOME is set. It can be reproduced on the byte-buddy project:

```
git clone https://github.com/raphw/byte-buddy
cd byte-buddy
git checkout byte-buddy-1.7.9
mvn org.apache.maven.plugins:maven-javadoc-plugin:3.0.0:javadoc
```

---

**Affects:** 3.0.0

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

How is the Maven script supposed to run when `JAVA_HOME` is not set?

**Comment by jira-importer** [comment]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** commented

Maven can run without `JAVA_HOME` using `which java` to locate the JRE:

https://github.com/apache/maven/blob/maven-3.5.2/apache-maven/src/bin/mvn#L92

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

That's correct,  I completely forgot it. Some components won't run. Do you expect to see a `BuildException`?

**Comment by jira-importer** [comment]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** commented

A BuildException wouldn't really help, that would still be a regression compared to the previous versions.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I agree, we should improve `JavadocUtil.getJavaHome` to check for `java.home` in case `JAVA_HOME` was not set. IDEs probably suffer with the same problem.
But can't be fixed yet due to an svn2git migration issue

**Comment by jira-importer** [comment]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** commented

Geronimo XBean is also affected by this issue. Curiously not all modules in the reactor are affected, the javadoc does build for the first modules but fails on the xbean-reflect module.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Does https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=shortlog;h=refs/heads/MJAVADOC-504 fix your issue?

**Comment by jira-importer** [comment]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** commented

I tried with the fix in commit 7505bb5 but it still triggers the same NullPointerException (javaHomeValue is null at line 1308). Maybe because the value of java.home is read from the environment variables instead of the system properties of the JVM?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I did a small change on the branch, can you verify that fix?

**Comment by jira-importer** [comment]
**[Emmanuel Bourg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebourg)** commented

It looks good, byte-buddy and xbean build fine now. Thank you for the fix!

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [79a6b4b69383d3bf5c342c12dc483bc3d4b82336](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=79a6b4b69383d3bf5c342c12dc483bc3d4b82336)
